illumina.py

```diff
@@ -400,44 +401,53 @@ def main_common_barcodes(args):
-        for key, value in element.items():
-            accumulator[key] = accumulator.get(key, 0) + value
-        return accumulator
+    reduce_db_file    = util.file.mkstempfname('sqlite_.db')
+    reduce_db     = util.file.CountDB(reduce_db_file)
+    reduce_db.start()
+    
+    barcodefile_tempfile_tuples = [(tile_barcode_file,util.file.mkstempfname('sqlite_.db')) for tile_barcode_file in tile_barcode_files]
 
-    with concurrent.futures.ProcessPoolExecutor(max_workers=len(tile_barcode_files)) as executor:
-        futures = [executor.submit(util.file.count_occurrences_in_tsv, filePath, include_noise=includeNoise) for filePath in tile_barcode_files]
+    # scatter tile-specific barcode files among workers to store barcode counts in SQLite
+    workers = util.misc.sanitize_thread_count(threads)
+    with concurrent.futures.ProcessPoolExecutor(max_workers=workers) as executor:
+        futures = [executor.submit(util.file.count_occurrences_in_tsv_sqlite_backed, tf, bf, include_noise=includeNoise) for bf,tf in barcodefile_tempfile_tuples]
         for future in concurrent.futures.as_completed(futures):
-            barcode_counts = sum_reducer(barcode_counts, future.result())
+            tmp_db, barcode_file = future.result()
+            log.debug("done reading barcodes from %s; adding to total...",barcode_file)
+            # gather and reduce counts from separate SQLite databases into one 
+            reduce_db.add_counts_from_other_db(tmp_db)
+            os.unlink(tmp_db)
 
-    # sort the counts, descending. Truncate the result if desired
-    log.info("sorting counts")
     illumina_reference = IlluminaIndexReference()
-    count_to_write = truncateToLength if truncateToLength else len(barcode_counts)
-    barcode_pairs_sorted_by_count = sorted(barcode_counts, key=barcode_counts.get, reverse=True)[:count_to_write]
 
-    mapped_counts = (   (k[:8], ",".join([x for x in illumina_reference.guess_index(k[:8], distance=1)] or ["Unknown"]), 
-                        k[8:], ",".join([x for x in illumina_reference.guess_index(k[8:], distance=1)] or ["Unknown"]), 
-                        barcode_counts[k]) 
-                    for k in barcode_pairs_sorted_by_count)
+    log.info("Number of barcodes seen %s",reduce_db.get_num_IDS())
 
     # write the barcodes and their corresponding counts
-    log.info("writing output")
     with open(outSummary, 'w') as tsvfile:
+        log.info("sorting counts...")
+        log.info("writing output...")
         writer = csv.writer(tsvfile, delimiter='\t')
         # write the header unless the user has specified not to do so
         if not omitHeader:
             writer.writerow(("Barcode1", "Likely_Index_Names1", "Barcode2", "Likely_Index_Names2", "Count"))
-        writer.writerows(mapped_counts)
+        chunk_size=100
+        for num_returned,rows in enumerate(util.misc.batch_iterator(itertools.islice(reduce_db.get_counts_descending(),0,truncateToLength),chunk_size)):
```

Comment:
Is batching necessary? The execute will fetch everything eagerly and the write should be batching writes behind the scenes.

Reply:
Thanks, yeah, it's not needed here. It is needed over in the SQLite class since each query can only handle so many sets of values.